chasm/tree.go

```diff
@@ -137,6 +152,378 @@ func NewEmptyTree(
+			continue
+		}
+
+		fieldN := fieldName(nodeValueT.Elem().Field(i))
+
+		// TODO: cache Data field in n and set it later to save on reflection.
+
+		switch genericTypePrefix(fieldT.Elem()) {
+		case chasmFieldTypePrefix:
+			internalV := fieldV.Elem().FieldByName(internalFieldName)
+			//nolint:revive // Internal field is guaranteed to be of type fieldInternal.
+			internal := internalV.Interface().(fieldInternal)
+			if internal.node != nil {
+				// This subcomponent already has a tree node. No need to create a new one.
+				continue
+			}
+
+			childN := newNode(n.nodeBase, n)
+			// Recursively set all sub-subcomponents.
+			if err := childN.setValue(internal.value, internal.fieldType); err != nil {
+				return err
+			}
+			n.children[fieldN] = childN
+			internal.node = childN
+			// TODO: this line can be remove if Internal becomes a *fieldInternal.
+			internalV.Set(reflect.ValueOf(internal))
+		case chasmCollectionTypePrefix:
+			panic("not implemented")
+		}
+	}
+	return nil
+}
+
+func (n *Node) setSerializedValue(
+	nodePath []string,
+	serializedNode *persistencespb.ChasmNode,
+) {
+	if len(nodePath) == 0 {
+		n.serializedValue = serializedNode
+		return
+	}
+
+	childName := nodePath[0]
+	childNode, ok := n.children[childName]
+	if !ok {
+		childNode = newNode(n.nodeBase, n)
+		n.children[childName] = childNode
+	}
+	childNode.setSerializedValue(nodePath[1:], serializedNode)
+}
+
+// TODO: this is called from CloseTransaction for every node in the tree.
+// serialize sets or updates serializedValue field of the node n with serialized value.
+func (n *Node) serialize() error {
+	switch n.serializedValue.Attributes.(type) {
+	case *persistencespb.ChasmNode_ComponentAttributes:
+		return n.serializeComponentNode()
+	case *persistencespb.ChasmNode_DataAttributes:
+		return n.serializeDataNode()
+	case *persistencespb.ChasmNode_CollectionAttributes:
+		panic("not implemented")
+	case *persistencespb.ChasmNode_PointerAttributes:
+		panic("not implemented")
+	default:
+		return serviceerror.NewInternal("unknown node type")
+	}
+}
+
+func (n *Node) serializeComponentNode() error {
+	nodeValueT := reflect.TypeOf(n.value)
+	nodeValueV := reflect.ValueOf(n.value)
+
+	protoMessageFound := false
+	for i := 0; i < nodeValueT.Elem().NumField(); i++ {
+		fieldV := nodeValueV.Elem().Field(i)
+		fieldT := fieldV.Type()
+		if !fieldT.AssignableTo(protoMessageT) {
+			continue
+		}
+
+		if fieldV.IsNil() {
+			continue
+		}
+
+		if protoMessageFound {
+			return serviceerror.NewInternal("only one proto field allowed in component")
+		}
+		protoMessageFound = true
+
+		blob, err := serialization.ProtoEncodeBlob(fieldV.Interface().(proto.Message), enumspb.ENCODING_TYPE_PROTO3)
+		if err != nil {
+			return err
+		}
+
+		n.serializedValue.GetComponentAttributes().Data = blob
+		n.serializedValue.GetComponentAttributes().Type = nodeValueT.Elem().String() // type name w/o *
+		if n.serializedValue.GetInitialVersionedTransition() == nil {
```

Comment:
initial versioned transition should only be set when the creating a new chasm node. It looks like you can do this as part of `setValue()`, though I am not 100% sure when that method will be called.

Then LastUpdateVersionedTransition is updated every time the node is mutated. Guess you can do it here.